diff -uNr a/arch/arm/boot/dts/qcom-ipq4019-ap.dk04.1-c1.dts b/arch/arm/boot/dts/qcom-ipq4019-ap.dk04.1-c1.dts
--- a/arch/arm/boot/dts/qcom-ipq4019-ap.dk04.1-c1.dts	2017-11-22 11:42:18.752133269 +0000
+++ b/arch/arm/boot/dts/qcom-ipq4019-ap.dk04.1-c1.dts	2017-12-16 08:50:45.327502468 +0000
@@ -272,6 +272,33 @@
 			status = "disabled";
 		};
 	};
+	gpio_keys {
+			compatible = "gpio-keys";
+
+			button@1 {
+				label = "reset";
+				linux,code = <KEY_RESTART>;
+				gpios = <&tlmm 18 GPIO_ACTIVE_LOW>;
+				linux,input-type = <1>;
+			};
+	};
+	leds-gpio {
+		compatible = "leds-gpio";
+
+		status_blue {
+			label = "status_blue";
+			gpios = <&tlmm 46 0>;
+		};
+
+		status_red {
+			label = "status_red";
+			gpios = <&tlmm 45 0>;
+		};
+		status_green {
+                        label = "status_green";
+                        gpios = <&tlmm 44 0>;
+                };
+	};
 };
 
 &spi_0 { /* BLSP1 QUP1 */
