diff -uNr a/arch/arm/boot/dts/qcom-ipq4019-ap.dk04.1-c1.dts b/arch/arm/boot/dts/qcom-ipq4019-ap.dk04.1-c1.dts
--- a/arch/arm/boot/dts/qcom-ipq4019-ap.dk04.1-c1.dts	2017-11-28 21:31:13.050678696 +0800
+++ b/arch/arm/boot/dts/qcom-ipq4019-ap.dk04.1-c1.dts	2017-11-28 21:32:07.006678039 +0800
@@ -330,7 +330,6 @@
 			partition5@e0000 {
 				label = "0:APPSBLENV"; /* uboot env*/
 				reg = <0x000e0000 0x00010000>;
-				read-only;
 			};
 			partition6@f0000 {
 				label = "0:APPSBL"; /* uboot */
@@ -340,7 +339,6 @@
 			partition7@170000 {
 				label = "0:ART";
 				reg = <0x00170000 0x00010000>;
-				read-only;
 			};
 			partition8@180000 {
 				label = "kernel";
